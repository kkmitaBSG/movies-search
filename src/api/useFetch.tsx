import { useEffect, useState } from "react";
import { useSearchParams } from "react-router-dom";
import messages from "../constants/messages";

export interface Items {
  id: string;
  title: string;
  overview?: string;
  original_language?: string;
}

interface UseFetchResponseType {
  status: string;
  results: Items[];
}

export enum LoadingStatus {
  Waiting = "waiting",
  Fetching = "fetching",
  Fetched = "fetched",
}

const useFetch = (searchValue = ""): UseFetchResponseType => {
  const [status, setStatus] = useState<LoadingStatus>(LoadingStatus.Waiting);
  const [results, setData] = useState<Items[]>([]);
  const [, setSearchParams] = useSearchParams();

  useEffect(() => {
    const fetchData = async () => {
      try {
        setStatus(LoadingStatus.Fetching);
        const response = await fetch(
          `https://api.themoviedb.org/3/search/movie?api_key=92b418e837b833be308bbfb1fb2aca1e&query=${searchValue}`
        );

        setSearchParams({ searchParam: searchValue });

        const data = await response.json();
        if (!response.ok) {
          throw response;
        }

        setData(data.results);
        setStatus(LoadingStatus.Fetched);
      } catch {
        console.log(messages.fetchResponse)
      }
    };

    fetchData();
  }, [searchValue, setSearchParams]);

  return { status, results };
};

export default useFetch;
