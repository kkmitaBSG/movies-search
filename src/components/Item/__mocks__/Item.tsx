import React, { ReactElement } from "react";
import { Movie } from "../../../store/types";

const MockedComponent = (movie: Movie): ReactElement => (
  <React.Fragment key={movie.id}>ItemMockedComponent</React.Fragment>
);

export default MockedComponent;
