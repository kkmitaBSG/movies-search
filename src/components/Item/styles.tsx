import styled from "styled-components";
import { NavLink } from "react-router-dom";

export const Item = styled(NavLink)<{ poster?: string }>`
  flex: 1 1 auto;
  margin: 10px;
  box-shadow: 0 0 10px #737373;
  border-radius: 10px;
  width: 250px;
  height: 450px;
  cursor: pointer;
  overflow: hidden;
  text-decoration: none;
  color: #FFF;

  ${({ poster }) => poster && `background-image: url(${poster});`}
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center center;
`;

export const AdditionalInfo = styled.div`
  opacity: 0;
  background-color: #00000090;
  height: 100%;
  padding: 50px 20px 0;
  border-radius: 10px;
  transition: opacity 0.3s ease-out;

  ${Item}:hover & {
    opacity: 1;
  }
`;

export const SubElem = styled.div`
  display: -webkit-box;
  -webkit-line-clamp: 13;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
`;

export const Name = styled.div`
  color: #ffbc9e;
  margin-bottom: 10px;
  font-size: 25px;
  font-weight: 500;
  text-align: center;
`;
