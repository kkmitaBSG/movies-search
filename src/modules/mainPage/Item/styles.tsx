import styled from "styled-components";

export const Item = styled.div`
  flex: 1 1 auto;
  margin: 10px;
  padding: 10px;
  box-shadow: 0 0 10px #aaa;
  border-radius: 10px;
`;

export const SubElem = styled.div``;

export const Name = styled(SubElem)`
  color: #0025aa;
  font-weight: 500;
`;

export const Desc = styled.span`
  font-size: 0.9em;
  color: #00ac84;
`;
