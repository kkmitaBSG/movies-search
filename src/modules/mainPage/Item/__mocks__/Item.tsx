import React, { ReactElement } from "react";
import { Items } from "../../../../api/useFetch";

const MockedComponent = (item: Items): ReactElement => (
  <React.Fragment key={item.id}>ItemMockedComponent</React.Fragment>
);

export default MockedComponent;
