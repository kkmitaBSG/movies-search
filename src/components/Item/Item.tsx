import React from "react";
import { Movie } from "../../store/types";
import * as S from "./styles";

const Item = React.memo(({ id, title, overview, poster }: Movie) => (
  <S.Item to={`details/${id}`} poster={poster}>
    <S.AdditionalInfo>
      <S.Name>{title}</S.Name>
      <S.SubElem>{overview}</S.SubElem>
    </S.AdditionalInfo>
  </S.Item>
));

export default Item;
