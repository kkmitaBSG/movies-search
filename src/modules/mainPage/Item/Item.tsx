import React from "react";
import { Items } from "../../../api/useFetch";
import messages from "../../../constants/messages";
import * as S from "./styles";

const Item = React.memo(({ title, overview, original_language }: Items) => (
    <S.Item>
      <S.Name>{title}</S.Name>
      {overview && (
        <S.SubElem>
          <S.Desc>{messages.overviewDesc}</S.Desc> {overview}
        </S.SubElem>
      )}
      {original_language && (
        <S.SubElem>
          <S.Desc>{messages.originalLanguageDesc}</S.Desc> {original_language}
        </S.SubElem>
      )}
    </S.Item>
  )
);

export default Item;
