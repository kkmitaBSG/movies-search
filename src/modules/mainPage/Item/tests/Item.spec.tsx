import Item from "../Item";
import renderer from "react-test-renderer";
import { Items } from "../../../../api/useFetch";

const defaultProps: Items = {
  id: "1",
  title: "item1Name",
  overview: "overview1",
  original_language: "en",
};

describe("Item", () => {
  test("with full filled props", () => {
    const component = renderer.create(<Item {...defaultProps} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test("with only required props", () => {
    const props = {
      id: "1",
      title: "item1Name",
    };

    const component = renderer.create(<Item {...props} />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
