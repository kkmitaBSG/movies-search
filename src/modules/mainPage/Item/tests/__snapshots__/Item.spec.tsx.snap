// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item with full filled props 1`] = `
<div
  className="sc-bcXHqe jDsvTA"
>
  <div
    className="sc-gswNZR sc-dkrFOg  gjmsdx"
  >
    item1Name
  </div>
  <div
    className="sc-gswNZR"
  >
    <span
      className="sc-hLBbgP jUjwA-D"
    >
      overview
    </span>
     
    overview1
  </div>
  <div
    className="sc-gswNZR"
  >
    <span
      className="sc-hLBbgP jUjwA-D"
    >
      original language
    </span>
     
    en
  </div>
</div>
`;

exports[`Item with only required props 1`] = `
<div
  className="sc-bcXHqe jDsvTA"
>
  <div
    className="sc-gswNZR sc-dkrFOg  gjmsdx"
  >
    item1Name
  </div>
</div>
`;
